src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala

```diff
@@ -6,68 +6,129 @@ import is.hail.utils._
-      // key.compare(elem)
-      i =>
-        if (i == numPartitions - 1)
-          -1 // key.compare(inf)
-        else
-          ordering.compare(pk, rangeBounds(i)))
-    part
+  // if outside bounds, throw error
+  // pk: Annotation[pkType]
+  def getPartitionPK(pk: Any): Int = {
+    assert(pkType.typeCheck(pk))
+    val part = rangeTree.queryValues(pkType.ordering, pk)
+    part match {
+      case Array(x) => x
+    }
   }
 
-  // return the smallest partition for which key <= max
+  // return the partition containing key
+  // if outside bounds, throw error
   // key: RegionValue[kType]
   def getPartition(key: Any): Int = {
     val keyrv = key.asInstanceOf[RegionValue]
+    val wpkrv = WritableRegionValue(pkType)
+    wpkrv.setSelect(kType, pkKFieldIdx, keyrv)
+    val pkUR = new UnsafeRow(pkType, wpkrv.value)
+
+    val part = rangeTree.queryValues(pkType.ordering, pkUR)
 
-    val part = BinarySearch.binarySearch(numPartitions,
-      // key.compare(elem)
-      i =>
-        if (i == numPartitions - 1)
-          -1 // key.compare(inf)
-        else
-          -pkKOrd.compare(rangeBounds.region, loadElement(i), keyrv))
-    part
+    part match {
+      case Array(x) => x
+    }
   }
 
   def withKType(newPartitionKey: Array[String], newKType: TStruct): OrderedRVDPartitioner = {
-    val newPart = new OrderedRVDPartitioner(numPartitions, newPartitionKey, newKType, rangeBounds)
+    val newPart = new OrderedRVDPartitioner(newPartitionKey, newKType, rangeBounds)
     assert(newPart.pkType == pkType)
     newPart
   }
+
+  def copy(numPartitions: Int = numPartitions, partitionKey: Array[String] = partitionKey,
+    kType: TStruct = kType, rangeBounds: UnsafeIndexedSeq = rangeBounds): OrderedRVDPartitioner = {
+    new OrderedRVDPartitioner(partitionKey, kType, rangeBounds)
+  }
+
+  def coalesceRangeBounds(newPartEnd: Array[Int]): OrderedRVDPartitioner = {
+    val newRangeBounds = UnsafeIndexedSeq(
+      rangeBoundsType,
+      (-1 +: newPartEnd.init).zip(newPartEnd).map { case (s, e) =>
+        val i1 = rangeBounds(s + 1).asInstanceOf[Interval]
+        val i2 = rangeBounds(e).asInstanceOf[Interval]
+        Interval(i1.start, i2.end, i1.includeStart, i2.includeEnd)
+      })
+    copy(numPartitions = newPartEnd.length, rangeBounds = newRangeBounds)
+  }
 }
 
 object OrderedRVDPartitioner {
   def empty(typ: OrderedRVDType): OrderedRVDPartitioner = {
-    new OrderedRVDPartitioner(0, typ.partitionKey, typ.kType, UnsafeIndexedSeq.empty(TArray(typ.pkType)))
+    new OrderedRVDPartitioner(typ.partitionKey, typ.kType, UnsafeIndexedSeq.empty(TArray(TInterval(typ.pkType))))
+  }
+
+  // takes npartitions + 1 points and returns npartitions intervals: [a,b], (b,c], (c,d], ... (i, j]
+  def makeRangeBoundIntervals(pType: Type, rangeBounds: Array[RegionValue]): UnsafeIndexedSeq =
+    makeRangeBoundIntervals(UnsafeIndexedSeq(TArray(pType), rangeBounds))
+
+  def makeRangeBoundIntervals(rangeBounds: UnsafeIndexedSeq): UnsafeIndexedSeq = {
```

Comment:
Only the array version is called now so I think you can merge these two.  Also, if you want to simplify your life a bit, you can just construct the Array[Interval] and call the UnsafeIndex constructor taking a type/array.